test/sdk/TopologyFixture.h

```diff
@@ -0,0 +1,207 @@
+        LOG(debug) << "$DDS_LOCATION:       " << std::getenv("DDS_LOCATION");
+        LOG(debug) << "$PATH:               " << std::string(std::getenv("PATH")).substr(0, 80) << " ...";
+    }
+
+    auto StartNewDDSSession() -> void {
+        mDDSSessionID = mDDSSession.create();
+        LOG(debug) << "DDSSession running?: " << mDDSSession.IsRunning();
+        LOG(debug) << "DDSSessionID:        " << mDDSSessionID;
+    }
+
+    auto SubmitDDSAgents() -> void {
+        // Prepare DDS submit request
+        LOG(debug) << "DDSTopologyFile:     " << mDDSTopologyFile;
+        LOG(debug) << "requiredNofAgents:   " << mRequiredNofAgents;
+        LOG(debug) << "DDSRMSPlugin:        " << mDDSRMSPlugin;
+        dds::tools_api::SSubmitRequestData submitInfo;
+        submitInfo.m_rms = mDDSRMSPlugin;
+        submitInfo.m_instances = mRequiredNofAgents;
+
+        // Submit DDS agents
+        // TODO enable aggregate initialization, e.g.
+        // ...SSubmitRequest::makeRequest({mRequiredNofAgents, mDDSRMSPlugin, "", ""});
+        auto submitRequest = dds::tools_api::SSubmitRequest::makeRequest(submitInfo);
+        submitRequest->setMessageCallback(
+            [](const dds::tools_api::SMessageResponseData& _message) { LOG(debug) << _message; });
+        submitRequest->setDoneCallback([&]() {
+            LOG(debug) << mRequiredNofAgents << " Agents submitted";
+            // TODO rename to unblockThread() or wakeThread(), it is confusing with
+            // CSession::shutdown
+            mDDSSession.stop();
+        });
+        mDDSSession.sendRequest<dds::tools_api::SSubmitRequest>(submitRequest);
+
+        // Wait for submission request to finish
+        mDDSSession.blockCurrentThread();
+    }
+
+    auto WaitForIdleDDSAgents(int required) -> void {
+        LOG(debug) << "WaitForIdleDDSAgents(" << required << ")";
+
+        // DDS Agent Info request
+        dds::tools_api::SAgentInfoRequestData agentInfoInfo;
+        auto agentInfoRequest = dds::tools_api::SAgentInfoRequest::makeRequest(agentInfoInfo);
+        agentInfoRequest->setResponseCallback(
+            [&](const dds::tools_api::SAgentInfoResponseData& _response) {
+                LOG(debug) << "active agents" << _response.m_activeAgentsCount;
+                // TODO rename to unblockThread() or wakeThread(), it is confusing with
+                // CSession::shutdown
+                mDDSSession.stop();
+            });
+        agentInfoRequest->setMessageCallback(
+            [](const dds::tools_api::SMessageResponseData& _message) { LOG(debug) << _message; });
+        agentInfoRequest->setDoneCallback([&]() {
+            // TODO rename to unblockThread() or wakeThread(), it is confusing with
+            // CSession::shutdown
+            mDDSSession.stop();
+        });
+        mDDSSession.sendRequest<dds::tools_api::SAgentInfoRequest>(agentInfoRequest);
+
+        // Wait for agent info request to finish
+        mDDSSession.blockCurrentThread();
```

Comment:
@AndreyLebedev Moved to next step and tried to do an agent info request (with FairRootGroup/DDS@21e8ad6) -> Program hangs, see `WaitForIdleDDSAgents` what I do:

```
╭─dklein@gamma ~/projects/FairMQ/build ‹ruby-2.6.0› ‹sdk_topology›
╰─➤  ctest -R SDK.Topology -VV
UpdateCTestConfiguration  from :/home/dklein/projects/FairMQ/build/DartConfiguration.tcl
Parse Config file:/home/dklein/projects/FairMQ/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/dklein/projects/FairMQ/build/DartConfiguration.tcl
Parse Config file:/home/dklein/projects/FairMQ/build/DartConfiguration.tcl
Test project /home/dklein/projects/FairMQ/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 137
    Start 137: SDK.Topology.Basic

137: Test command: /home/dklein/projects/FairMQ/build/test/testsuite_SDK "--gtest_filter=Topology.Basic" "--gtest_also_run_disabled_tests"
137: Test timeout computed to be: 5
137: Note: Google Test filter = Topology.Basic
137: [==========] Running 1 test from 1 test case.
137: [----------] Global test environment set-up.
137: [----------] 1 test from Topology
137: [ RUN      ] Topology.Basic
137: [20:48:25.299865][INFO] Load DDS environment
137: [20:48:25.300010][DEBUG] >>>
137: [20:48:25.300023][DEBUG]  DDS_CFG=`dds-user-defaults --ignore-default-sid -p`
137: [20:48:25.300031][DEBUG]  if [ -z "$DDS_CFG" ]; then
137: [20:48:25.300037][DEBUG]    dds-user-defaults --ignore-default-sid -d -c "$HOME/.DDS/DDS.cfg"
137: [20:48:25.300043][DEBUG]  fi
137: [20:48:25.300048][DEBUG] >>> 0
137: [20:48:25.304743][DEBUG] $HOME:               /home/dklein/projects/FairMQ/build/test
137: [20:48:25.304761][DEBUG] $DDS_LOCATION:       /home/dklein/projects/DDS/build/install
137: [20:48:25.304768][DEBUG] $PATH:               /home/dklein/projects/DDS/build/install/bin:/home/dklein/projects/DDS/build/inst ...
137: [20:48:25.304787][DEBUG] DDSSession running?: 0
137: [20:48:25.306030][INFO] Start new DDS session
137: [20:48:25.382984][DEBUG] DDSSession running?: 1
137: [20:48:25.383095][DEBUG] DDSSessionID:        22f5da01-e54b-41c1-8b83-cdc1bb46f19b
137: [20:48:25.383105][INFO] Submit DDS agents
137: [20:48:25.383111][DEBUG] DDSTopologyFile:     /home/dklein/projects/FairMQ/test/sdk/test_topo.xml
137: [20:48:25.383117][DEBUG] requiredNofAgents:   12
137: [20:48:25.383123][DEBUG] DDSRMSPlugin:        localhost
137: [20:48:25.386034][DEBUG] <info> Creating new worker package...
137: [20:48:25.534696][DEBUG] <info> RMS plug-in: /home/dklein/projects/DDS/build/install/plugins/dds-submit-localhost/dds-submit-localhost
137: [20:48:25.535117][DEBUG] <info> Initializing RMS plug-in...
137: [20:48:25.553456][DEBUG] <info> RMS plug-in is online. Startup time: 18ms.
137: [20:48:25.555182][DEBUG] <info> Plug-in: Will use the local host to deploy 12 agents
137: [20:48:25.595607][DEBUG] <info> Plug-in: Using '/tmp/22f5da01-e54b-41c1-8b83-cdc1bb46f19b/dds_2019-07-12-20-48-25-554' to spawn agents
137: [20:48:25.596232][DEBUG] <info> Plug-in: Starting DDSScout in '/tmp/22f5da01-e54b-41c1-8b83-cdc1bb46f19b/dds_2019-07-12-20-48-25-554/wn'
137: [20:48:25.596801][DEBUG] <info> Plug-in: DDS agents have been submitted
137: [20:48:25.597202][DEBUG] <info> Plug-in: Checking status of agents...
137: [20:48:25.958714][DEBUG] <info> Plug-in: All agents have been started successfully
137: [20:48:25.959238][DEBUG] <info> Plug-in: Validating...
137: [20:48:26.164537][DEBUG] <info> Plug-in: All agents have been validated successfully
137: [20:48:26.166471][DEBUG] 12 Agents submitted
137: [20:48:26.167675][INFO] Wait for 12 idle DDS agents
137: [20:48:26.167891][DEBUG] WaitForIdleDDSAgents(12)
1/1 Test #137: SDK.Topology.Basic ...............***Timeout   30.03 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   5.04 sec

The following tests FAILED:
        137 - SDK.Topology.Basic (Timeout)
Errors while running CTest
```

Reply:
@AndreyLebedev I think I understand the issue here. `CSession::BlockCurrentThread()` is not really resumable at the moment, because unblocking must be done via `CSession::stop()` which in addition also call `CSession::unsubscribe()` which I believe causes the hanging program. We need to resolve FairRootGroup/DDS#225 first which should fix this problem here.

Reply:
I worked around `CSession::blockCurrentThread()` (I do not use it for now), so the issue is not blocking anything on our side any more.